hail/python/hail/experimental/db.py

```diff
@@ -1,60 +1,284 @@
+
+    def select(self, *args, **kwargs):
+        return RowLens(self.mt.select_rows(*args, **kwargs))
+
+    def explode(self, *args, **kwargs):
+        return RowLens(self.mt.explode_rows(*args, **kwargs))
+
+    def group_by(self, *args, **kwargs):
+        return GroupedRowLens(self.mt.group_rows_by(*args, **kwargs))
+
+    def __getitem__(self, *args, **kwargs):
+        return self.mt.__getitem__(*args, **kwargs)
+
+    def index(self, *args, **kwargs):
+        return self.mt.rows().index(*args, **kwargs)
+
+    def unlens(self):
+        return self.mt
+
+
+class GroupedRowLens:
+    def __init__(self, mt):
+        assert isinstance(mt, hl.GroupedMatrixTable)
+        self.mt = mt
+
+    def aggregate(self, *args, **kwargs):
+        return RowLens(self.mt.aggregate_rows(*args, **kwargs).result())
+
+
+class NoLens:
+    def __init__(self, t):
+        assert isinstance(t, hl.Table)
+        self.t = t
+        self.key = t.key
+
+    def annotate(self, *args, **kwargs):
+        return NoLens(self.t.annotate(*args, **kwargs))
+
+    def drop(self, *args, **kwargs):
+        return NoLens(self.t.drop(*args, **kwargs))
+
+    def select(self, *args, **kwargs):
+        return NoLens(self.t.select(*args, **kwargs))
+
+    def explode(self, *args, **kwargs):
+        return NoLens(self.t.explode(*args, **kwargs))
+
+    def group_by(self, *args, **kwargs):
+        return GroupedNoLens(self.t.group_by(*args, **kwargs))
+
+    def __getitem__(self, *args, **kwargs):
+        return self.t.__getitem__(*args, **kwargs)
+
+    def index(self, *args, **kwargs):
+        return self.t.index(*args, **kwargs)
+
+    def unlens(self):
+        return self.t
+
+
+class GroupedNoLens:
+    def __init__(self, t):
+        assert isinstance(t, hl.GroupedTable)
+        self.t = t
+
+    def aggregate(self, *args, **kwargs):
+        return NoLens(self.t.aggregate(*args, **kwargs))
+
+
 class DB:
-    
-    _annotation_dataset_urls = None
-    
+    """An annotation database instance.
+
+    This class facilitates the annotation of genetic datasets with variant
+    annotations. It accepts either an HTTP(S) URL to an Annotation DB
+    configuration or a python :obj:`dict` describing an Annotation DB
+    configuration.
+
+    Examples
+    --------
+    Create an annotation database connecting to the default Hail Annotation DB:
+
+    >>> db = hl.experimental.DB()
+    >>> mt = db.annotate_rows_db(mt, 'gnomad_lof_metrics') # doctest: +SKIP
+    """
+
+    _valid_key_properties = {'gene', 'unique'}
+
+    def __init__(self,
+                 url=None,
+                 json=None):
+        if json is not None and url is not None:
+            raise ValueError(f'Only specify one of the parameters url and json, '
+                             f'received: url={url} and json={json}')
+        if json is None:
+            if url is None:
+                url = ('https://www.googleapis.com/storage/v1/b/hail-common/o/annotationdb%2f' +
```

Comment:
I'd prefer to put this JSON file in the Python package. It seems like the pro of putting it on GS is that we can edit it without a deploy, but I'd want to version that in our monorepo anyhow.

Reply:
I am also conflicted. This file is versioned (it's `annotation_db.json`).

Including the JSON file in the python package seems to require that these datasets continue to exist at the given URLs as long as `0.2.x` is supported. I think I'm OK with that if we think a bit about the ramifications of that and pick a supportable location.

Reply:
I think we can only reason about the system if we can guarantee that the version in the repo is the same as the version that the build is pointing at on the cloud, and if that's true, we might as well just keep the file in the package.

The alternative to requiring that datasets continue to exist is updating every json file we've ever used. That seems harder.

I'm also generally OK with saying "we broke the annotation DB, please update to latest hail"

Reply:
I aspire to replication of versioned Hail scripts, including those that read the annotation database, at least within some "lifetime" of each major version.

I'll think about where these datasets should live so I feel confident checking the file in.